idseq_dag/steps/generate_coverage_stats.py

```diff
@@ -48,85 +66,105 @@ def run(self):
-                with Pool() as pool:
-                    chunk_results = pool.map(fn, chunked_references)
-        contig2coverage = {}
-        for chunk_result in chunk_results:
-            # check if there are duplicated keys
-            for key in chunk_result.keys():
-                if key in contig2coverage:
-                    raise Exception(f"Key {key} already present in contig2coverage")
-            contig2coverage.update(chunk_result)
-        return contig2coverage
+    def calc_contig2coverage(bam_filename):
+        # PySAM pileup is CPU-intenstive.  Each CPU core is assigned a slice of the input bam file on which to perform pileup.  The slice contigs are selected by slice_idx modulo num_slices.  Each slice gets its own pair of temporary output files, one in CSV format and one in JSON.  In the end, these slice outputs are concatenated.  This is a similar pattern to run_lzw.
+        num_physical_cpu = (cpu_count() + 1) // 2
+        num_slices = num_physical_cpu
+        output_csv_filenames = [f"tmp_slice_{num_slices}_{slice}.csv" for slice in range(num_slices + 1)]
+        output_json_filenames = [f"tmp_slice_{num_slices}_{slice}.json" for slice in range(num_slices + 1)]
+        @run_in_subprocess
+        def compute_slice(slice_idx):
+            with open(output_csv_filenames[slice_idx], "w") as output_csv, \
+                open(output_json_filenames[slice_idx], "w") as output_json, \
+                pysam.AlignmentFile(bam_filename, "rb") as input_bam:
+                for contig_idx, contig_name in enumerate(input_bam.references):
+                    if contig_idx % num_slices == slice_idx:
+                        PipelineStepGenerateCoverageStats._process_contig(input_bam, output_csv, output_json, contig_name)
+        # Compute pileup for each slice
+        with LongRunningCodeSection("PipelineStepGenerateCoverageStats.calc_contig2coverage.mt_map"):
+            mt_map(compute_slice, range(num_slices))
+        # Output CSV headers
+        with output_csv_filenames[-1] as ocsv:
+            ocsv.write(",".join(COVERAGE_STATS_SCHEMA))
+            ocsv.write("\n")
+        # Output JSON dict open paren
+        with output_json_filenames[-1] as ojson:
+            ojson.write("{")
+        # Collate CSV slices
+        command.execute(
+            command_patterns.ShellScriptCommand(
+                script=r'''cat "${individual_slice_outputs[@]}" > "${collated_csv}";''',
+                named_args={
+                    'collated_csv': output_csv_filenames[-1],
+                    'individual_slice_outputs': output_csv_filenames[:-1]
+                }
+            )
+        )
+        for tfn in output_csv_filenames[:-1]:
+            os.remove(tfn)
+        # Collate JSON slices, replacing final ", " with "}"
+        command.execute(
+            command_patterns.ShellScriptCommand(
+                script=r'''cat "${individual_slice_outputs[@]}" | sed 's=, $=}=' > "${collated_json}";''',
```

Comment:
I wanted to double check if I understood this part correctly. Do you want to replace the colon only on the last line of this file?

I tested this `sed` command:

```
function my_test() {
     # sample file
     (for X in {1..2}; do echo '{"line": '${X}'}, '; done) > test1.txt
     (for X in {3..4}; do echo '{"line": '${X}'}, '; done) > test2.txt
 
     echo "***original files***:"
     cat test1.txt test2.txt
 
     # reproducing that sed instruction
     cat test1.txt test2.txt | sed 's=, $=}=' > "transformed.txt"
 
     echo
     echo "***dest file***:"
     cat transformed.txt
}
my_test
```

That produces this output:
```
***original files***:
{"line": 1}, 
{"line": 2}, 
{"line": 3}, 
{"line": 4}, 

***dest file***:
{"line": 1}}
{"line": 2}}
{"line": 3}}
{"line": 4}}
```

I thought maybe you only wanted line 4 to be replaced. Is that the case?

Reply:
There is only a single line in that file.

Reply:
Oh I see. Well, in this case it will work as expected.

Just in case we want to be more specific, adding a `$` sign in front of that pattern (i.e. `$s=, $=}=`)will make it work as expected even when using more than one line:
```
cat test1.txt test2.txt | sed '$s=, $=}=' > "transformed.txt"
```

would produce this output:

```
***dest file***:
{"line": 1}, 
{"line": 2}, 
{"line": 3}, 
{"line": 4}}
```